### DISCUSSION TITLE
레벨4 테스트 개선에 대한 의견

### DISCUSSION THREAD

**Discussion by Yboyu0u** [woowacourse/prolog#1017] [Ideas]
이슈 번호: https://github.com/woowacourse/prolog/issues/1004

다들 팀프로젝트에서 테스트 전략을 어떻게 가져가고 있는지 공유해 주세요.

**Comment by leo0842** [comment]
어케 적어야 할지 감이 잘 안잡혀요 일단 적어볼게용

## 인수 테스트
- rest assured 를 사용
- 예외 테스트는 공통된 예외(인증, 인가 등)를 제외한 api 의 특수 예외에만 적용

## 컨트롤러 테스트
- 슬라이스 테스트
- mockMvc 활용
- rest docs 용도로 주요 사용

## 서비스 레이어 테스트
- 처음에는 모킹하여 슬라이스 테스트로 하다가 실제 데이터로 하는 클래식?? 테스트로 변경

## 레포지토리 레이어 테스트
- 슬라이스 테스트
- DataJpaTest 활용

**Comment by wishoon** [comment]
## 인수 테스트
* Rest Assured를 사용
* 시나리오 테스트 위주로 작성하며, 사용자 입장에서 프로그램을 사용하는 과정을 그리고 테스트로 작성함
* 단순 실패 테스트 작성이 아니라 의도된 실패 시나리오도 작성되면 좋을 것 같음

## 컨트롤러 테스트
* MockMvc를 사용함
* Rest Docs를 사용하기 위한 용도로 사용하며, `@Valid` 체크 용도로도 일부 사용함

## 서비스 테스트
* `@SpringBootTest`를 사용함
* 팀 프로젝트에는 `@SpringBootTest`를 사용하고 있지만, 개인적으로 Mocking 하는 것을 선호함 (슬라이스 테스트)
* 도메인 테스트에서 수행하지 못하는 예외 테스트를 검증함
* 만약, `@SpringBootTest`를 사용한다면, 컨텍스트가 띄워지는 시간을 줄인다면 충분히 활용할 수 있을거라고 생각

## 도메인 테스트 && 레포지토리 테스트
* 순수 자바 로직 테스트와 `@DataJapTest` 사용 (단위 테스트)
* 테스트 중 도메인 테스트의 비중을 높게 가져감

**Comment by EastHShin** [comment]
## 인수테스트
- RestAssured 사용
- 시나리오를 도출해 BDD 방식으로 테스트 작성
## 컨트롤러 테스트
- RestAssuredMockMvc 사용
    - RestAssured와 비슷한 문법으로 사용 가능
- Rest docs 작성 용도로 테스트 작성
## 서비스 테스트
- `@SpringBootTest` 를 사용한 통합테스트
- 사용자 시나리오가 아닌, 메서드 단위의 기능 정상 동작과 예외 검증을 위한 용도
## 도메인 테스트
- 객체의 행위를 검증하는 도메인 단위 테스트
